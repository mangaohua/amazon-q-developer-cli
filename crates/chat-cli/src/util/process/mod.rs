pub use sysinfo::Pid;

#[cfg(target_os = "windows")]
mod windows;
#[cfg(target_os = "windows")]
pub use windows::*;

#[cfg(not(windows))]
mod unix;
#[cfg(not(windows))]
pub use unix::*;
