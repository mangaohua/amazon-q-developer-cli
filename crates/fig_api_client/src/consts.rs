use aws_config::Region;

// Endpoint constants
pub const PROD_CODEWHISPERER_ENDPOINT_URL: &str = "https://codewhisperer.us-east-1.amazonaws.com";
pub const PROD_CODEWHISPERER_ENDPOINT_REGION: Region = Region::from_static("us-east-1");

pub const PROD_Q_ENDPOINT_URL: &str = "https://q.us-east-1.amazonaws.com";
pub const PROD_Q_ENDPOINT_REGION: Region = Region::from_static("us-east-1");

// Opt out constants
pub const SHARE_CODEWHISPERER_CONTENT_SETTINGS_KEY: &str = "codeWhisperer.shareCodeWhispererContentWithAWS";
pub const X_AMZN_CODEWHISPERER_OPT_OUT_HEADER: &str = "x-amzn-codewhisperer-optout";

// Session ID constants
pub const X_AMZN_SESSIONID_HEADER: &str = "x-amzn-sessionid";
