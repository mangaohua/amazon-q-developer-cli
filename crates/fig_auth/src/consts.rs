use aws_types::region::Region;

pub(crate) const CLIENT_NAME: &str = "Amazon Q Developer for command line";

pub(crate) const OIDC_BUILDER_ID_REGION: Region = Region::from_static("us-east-1");

/// The scopes requested for OIDC
///
/// Do not include `sso:account:access`, these permissions are not needed and were
/// previously included
pub(crate) const SCOPES: &[&str] = &[
    "codewhisperer:completions",
    "codewhisperer:analysis",
    "codewhisperer:conversations",
    // "codewhisperer:taskassist",
    // "codewhisperer:transformations",
];

pub(crate) const CLIENT_TYPE: &str = "public";

// The start URL for public builder ID users
pub const START_URL: &str = "https://view.awsapps.com/start";

// The start URL for internal amzn users
pub const AMZN_START_URL: &str = "https://amzn.awsapps.com/start";

pub(crate) const DEVICE_GRANT_TYPE: &str = "urn:ietf:params:oauth:grant-type:device_code";
pub(crate) const REFRESH_GRANT_TYPE: &str = "refresh_token";
