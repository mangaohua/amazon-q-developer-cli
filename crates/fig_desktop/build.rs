use std::path::PathBuf;
use std::{
    env,
    fs,
};

use image::imageops::FilterType;

fn main() {
    build_assets();
}

fn build_assets() {
    resize_directory(
        "icons/autocomplete",
        "AUTOCOMPLETE_ICONS_PROCESSED",
        32,
        32,
        FilterType::Lanczos3,
    );
}

fn resize_directory(name: &str, var: &str, width: u32, height: u32, filter: FilterType) {
    println!("cargo:rerun-if-changed={name}");
    let source = env::current_dir().unwrap().join(name);
    let target = PathBuf::from(env::var("OUT_DIR").unwrap()).join(name);
    fs::create_dir_all(&target).expect("Failed creating assets folder");

    for (name, path) in source
        .read_dir()
        .expect("Failed reading nested assets directory")
        .map(Result::unwrap)
        .map(|x| (x.file_name(), x.path()))
    {
        let asset = image::open(path.clone())
            .unwrap_or_else(|err| panic!("Failed reading image at path {}: {:?}", &path.to_string_lossy(), err));
        let asset = asset.resize_exact(width, height, filter);
        asset.save(target.join(name)).expect("Failed writing image");
    }

    println!("cargo:rustc-env={}={}", var, target.to_string_lossy());
}
