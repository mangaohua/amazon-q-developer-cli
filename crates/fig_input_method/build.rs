use std::fs;

use apple_bundle::prelude::{
    InfoPlist as AppleInfoPlist,
    *,
};
use serde::{
    Deserialize,
    Serialize,
};

#[derive(Debug, Serialize, Deserialize)]
struct CargoToml {
    package: CargoTomlPackage,
}

#[derive(Debug, Serialize, Deserialize)]
struct CargoTomlPackage {
    metadata: CargoTomlPackageMetadata,
}

#[derive(Debug, Serialize, Deserialize)]
struct CargoTomlPackageMetadata {
    bundle: BundleConfig,
}

#[derive(Debug, Serialize, Deserialize)]
struct BundleConfig {
    #[serde(rename(serialize = "CFBundleIdentifier"))]
    bundle_identifier: String,
    #[serde(rename(serialize = "CFBundleName"))]
    bundle_name: String,
    // #[serde(rename(serialize = "CFBundleShortVersionString"))]
    // version: String,
    #[serde(flatten)]
    input_method: InputMethod,
}

#[derive(Debug, Serialize, Deserialize)]
struct InfoPlist {
    #[serde(flatten)]
    apple: AppleInfoPlist,
    #[serde(flatten)]
    input_method: InputMethod,
}

#[derive(Debug, Serialize, Deserialize)]
struct InputMethod {
    #[serde(
        rename(serialize = "InputMethodConnectionName"),
        serialize_with = "serialize_option",
        skip_serializing_if = "Option::is_none"
    )]
    input_method_connection_name: Option<String>,
    #[serde(
        rename(serialize = "InputMethodServerControllerClass"),
        default = "input_method_server_controller_class_default"
    )]
    input_method_server_controller_class: String,
    #[serde(rename(serialize = "TISInputSourceID"))]
    input_source_identifier: String,
    #[serde(rename(serialize = "InputMethodType"))]
    input_method_type: String,
    #[serde(rename(serialize = "ComponentInvisibleInSystemUI"))]
    invisible_in_system_ui: bool,
    #[serde(
        rename(serialize = "tsInputMethodIconFileKey"),
        default = "input_method_file_icon_key_default"
    )]
    input_method_file_icon_key: String,
    #[serde(rename(serialize = "TISIntendedLanguage"), default = "intended_language_default")]
    intended_language: String,
}

fn serialize_option<S, T>(value: &Option<T>, ser: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
    T: Serialize,
{
    value
        .as_ref()
        .expect(r#"`serialize_option` must be used with `skip_serializing_if = "Option::is_none"`"#)
        .serialize(ser)
}

fn input_method_server_controller_class_default() -> String {
    "RustInputMethodServerController".to_string()
}

fn input_method_file_icon_key_default() -> String {
    "AppIcon".to_string()
}

fn intended_language_default() -> String {
    "en".to_string()
}

fn main() {
    // println!("cargo:warning=Running build.rs");
    // Tell Cargo that if the given file changes, to rerun this build script.
    println!("cargo:rerun-if-changed=build.rs");
    println!("cargo:rerun-if-changed=Cargo.toml");

    // Read Cargo.toml and load [bundle]
    let config = fs::read_to_string("Cargo.toml").unwrap();

    let manifest = toml::from_str::<CargoToml>(&config).unwrap();
    let mut bundle = manifest.package.metadata.bundle;

    // Set InputMethodServerControllerClass env var which is used by Input Method at compile time
    println!(
        "cargo:rustc-env=InputMethodServerControllerClass={}",
        bundle.input_method.input_method_server_controller_class
    );

    // Derieve InputMethodConnectionName from bundle identifier
    let connection_name = match bundle.input_method.input_method_connection_name {
        Some(name) => name,
        None => format!("{}_Connection", bundle.bundle_identifier),
    };

    bundle.input_method.input_method_connection_name = Some(connection_name.clone());

    println!("cargo:rustc-env=InputMethodConnectionName={connection_name}");

    let properties = InfoPlist {
        apple: AppleInfoPlist {
            localization: Localization {
                bundle_development_region: Some("en".to_owned()),
                ..Default::default()
            },
            launch: Launch {
                bundle_executable: Some(env!("CARGO_PKG_NAME").to_owned()),
                ..Default::default()
            },
            identification: Identification {
                bundle_identifier: bundle.bundle_identifier,
                ..Default::default()
            },
            bundle_version: BundleVersion {
                bundle_version: Some(env!("CARGO_PKG_VERSION").to_string()),
                bundle_info_dictionary_version: Some("6.0".to_string()),
                bundle_short_version_string: Some(env!("CARGO_PKG_VERSION").to_string()),
                human_readable_copyright: Some(format!(
                    "Copyright Â© 2022 {} All rights reserved.",
                    env!("CARGO_PKG_AUTHORS")
                )),
            },
            naming: Naming {
                bundle_name: Some(bundle.bundle_name),
                ..Default::default()
            },
            categorization: Categorization {
                bundle_package_type: Some("APPL".to_owned()),
                ..Default::default()
            },
            background_execution: BackgroundExecution {
                background_only: Some(true),
                ..Default::default()
            },
            termination: Termination {
                supports_sudden_termination: Some(true),
                ..Default::default()
            },
            operating_system_version: OperatingSystemVersion {
                minimum_system_version: Some("10.13".to_string()),
                ..Default::default()
            },
            ..Default::default()
        },
        input_method: bundle.input_method,
    };

    // println!("cargo:warning=Regenerating Info.plist");

    // Create Info.plist file
    let file = std::fs::File::create("Info.plist").unwrap();
    plist::to_writer_xml(file, &properties).unwrap();
}
