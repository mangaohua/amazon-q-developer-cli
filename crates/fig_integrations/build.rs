const CODEX_FOLDER: &str = "src/shell/inline_shell_completion";

// The order here is very specific, do no edit without understanding the implications
const CODEX_FILES: &[&str] = &[
    "guard_start.zsh",
    "LICENSE",
    "config.zsh",
    "util.zsh",
    "bind.zsh",
    "highlight.zsh",
    "widgets.zsh",
    "strategies/inline.zsh",
    "strategies/completion.zsh",
    "strategies/history.zsh",
    "strategies/match_prev_cmd.zsh",
    "fetch.zsh",
    "async.zsh",
    "start.zsh",
    "guard_end.zsh",
];

fn main() {
    println!("cargo:rerun-if-changed=build.rs");

    let out_dir = std::env::var("OUT_DIR").unwrap();
    let out_dir = std::path::Path::new(&out_dir);

    let mut inline_shell_completion = String::new();
    for file in CODEX_FILES {
        let path = std::path::Path::new(CODEX_FOLDER).join(file);
        println!("cargo:rerun-if-changed={}", path.display());
        inline_shell_completion.push_str(&std::fs::read_to_string(path).unwrap());
    }
    std::fs::write(out_dir.join("inline_shell_completion.zsh"), inline_shell_completion).unwrap();
}
