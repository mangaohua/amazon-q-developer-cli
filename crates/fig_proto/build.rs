use std::io::Result;
use std::path::PathBuf;
use std::process::Command;

#[allow(dead_code)]
enum Version {
    V1([u32; 3]),
    V2([u32; 2]),
}

/// Try to find the version of protoc installed on the system.
fn protoc_version() -> Option<Version> {
    let output = std::process::Command::new("protoc").arg("--version").output().ok()?;
    let version = String::from_utf8(output.stdout).ok()?;
    eprintln!("protoc version: {version:?}");

    let version = version.trim();
    eprintln!("version: {version:?}");
    let version = version.split(' ').last().expect("No version");
    let version = version.split('.').collect::<Vec<_>>();
    let version = version
        .iter()
        .map(|s| s.parse::<u32>().ok())
        .collect::<Option<Vec<_>>>()?;
    match version.len() {
        3 => Some(Version::V1([version[0], version[1], version[2]])),
        2 => Some(Version::V2([version[0], version[1]])),
        _ => None,
    }
}

fn download_protoc() {
    let protoc_version = "26.1";

    let tmp_folder = tempfile::tempdir().unwrap();

    let os = match std::env::consts::OS {
        "linux" => "linux",
        "macos" => "osx",
        os => panic!("Unsupported os: {os}"),
    };

    let arch = match std::env::consts::ARCH {
        "x86_64" => "x86_64",
        "aarch64" => "aarch_64",
        arch => panic!("Unsupported arch: {arch}"),
    };

    let checksum = match (os, arch) {
        ("linux", "x86_64") => "a7be2928c0454f132c599e25b79b7ad1b57663f2337d7f7e468a1d59b98ec1b0",
        ("linux", "aarch_64") => "64a3b3b5f7dac0c8f9cf1cb85b2b1a237eb628644f6bcb0fb8f23db6e0d66181",
        ("osx", "x86_64") => "febd8821c3a2a23f72f4641471e0ab6486f4fb07b68111490a27a31681465b3c",
        ("osx", "aarch_64") => "26a29befa8891ecc48809958c909d284f2b9539a2eb47f22cadc631fe6abe8fd",
        _ => unreachable!(),
    };

    let mut download_command = Command::new("curl");
    download_command
        .arg("-Lf")
        .arg(format!(
            "https://github.com/protocolbuffers/protobuf/releases/download/v{protoc_version}/protoc-{protoc_version}-{os}-{arch}.zip"
        ))
        .arg("-o")
        .arg(tmp_folder.path().join("protoc.zip"));
    assert!(download_command.spawn().unwrap().wait().unwrap().success());

    let mut checksum_command = Command::new("sha256sum");
    checksum_command.arg(tmp_folder.path().join("protoc.zip"));
    let checksum_output = checksum_command.output().unwrap();
    let checksum_output = String::from_utf8(checksum_output.stdout).unwrap();

    eprintln!("checksum: {checksum_output:?}");
    assert!(checksum_output.starts_with(checksum));

    let mut unzip_command = Command::new("unzip");
    unzip_command
        .arg("-o")
        .arg(tmp_folder.path().join("protoc.zip"))
        .current_dir(tmp_folder.path());
    assert!(unzip_command.spawn().unwrap().wait().unwrap().success());

    let out_bin = PathBuf::from(std::env::var("OUT_DIR").unwrap()).join("protoc");

    let mut mv = Command::new("mv");
    mv.arg(tmp_folder.path().join("bin/protoc")).arg(&out_bin);
    assert!(mv.spawn().unwrap().wait().unwrap().success());

    std::env::set_var("PROTOC", out_bin);
}

fn main() -> Result<()> {
    println!("cargo:rerun-if-changed=build.rs");

    let proto_files = std::fs::read_dir("../../proto")?
        .filter_map(|entry| entry.ok())
        .filter(|entry| entry.file_type().is_ok_and(|t| t.is_file()))
        .filter(|entry| entry.path().extension().is_some_and(|ext| ext == "proto"))
        .map(|entry| entry.path())
        .collect::<Vec<_>>();

    for file in &proto_files {
        println!("cargo:rerun-if-changed={}", file.display());
    }

    // --experimental_allow_proto3_optional is supported only on version of protoc >= 3.12
    // if the version of the system protoc is too old, we must panic
    match protoc_version() {
        Some(Version::V1([0..=2, _, _] | [3, 0..=11, _])) => download_protoc(),
        Some(Version::V1(_) | Version::V2(_)) => {},
        None => download_protoc(),
    };

    let mut config = prost_build::Config::new();

    config.protoc_arg("--experimental_allow_proto3_optional");

    #[cfg(feature = "arbitrary")]
    config.type_attribute(
        ".",
        "#[cfg_attr(feature = \"arbitrary\", derive(arbitrary::Arbitrary))]",
    );

    config.extern_path(".fig_common.Empty", "()");

    prost_reflect_build::Builder::new()
        .file_descriptor_set_path(PathBuf::from(std::env::var("OUT_DIR").unwrap()).join("file_descriptor_set.bin"))
        .descriptor_pool("crate::DESCRIPTOR_POOL")
        .compile_protos_with_config(config, &proto_files, &["../../proto"])?;

    Ok(())
}
