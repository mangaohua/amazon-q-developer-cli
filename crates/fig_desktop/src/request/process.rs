use std::time::Duration;

use fig_desktop_api::requests::Error;
use fig_proto::fig::server_originated_message::Submessage as ServerOriginatedSubMessage;
use fig_proto::fig::{
    EnvironmentVariable,
    PseudoterminalExecuteRequest,
    PseudoterminalExecuteResponse,
    RunProcessRequest,
    RunProcessResponse,
};
use fig_proto::remote::hostbound;
use fig_remote_ipc::figterm::{
    FigtermCommand,
    FigtermSessionId,
    FigtermState,
};
use fig_util::env_var::{
    PROCESS_LAUNCHED_BY_Q,
    Q_TERM,
};
use tokio::process::Command;
use tokio::time::timeout;
use tracing::{
    debug,
    warn,
};

use super::{
    RequestResult,
    RequestResultImpl,
};
use crate::platform::PlatformState;

fn shell_args(shell_path: &str) -> &'static [&'static str] {
    let (_, shell_name) = shell_path.rsplit_once(['/', '\\']).unwrap_or(("", shell_path));
    match shell_name {
        "bash" | "bash.exe" => &["--norc", "--noprofile", "-c"],
        "zsh" | "zsh.exe" => &["--norcs", "-c"],
        "fish" | "fish.exe" => &["--no-config", "-c"],
        _ => {
            warn!(%shell_name, "unknown shell");
            &[]
        },
    }
}

fn set_fig_vars(cmd: &mut Command) {
    cmd.env(Q_TERM, env!("CARGO_PKG_VERSION"));
    cmd.env(PROCESS_LAUNCHED_BY_Q, "1");

    cmd.env("HISTFILE", "");
    cmd.env("HISTCONTROL", "ignoreboth");
    cmd.env("TERM", "xterm-256color");
}

pub async fn execute(request: PseudoterminalExecuteRequest, figterm_state: &FigtermState) -> RequestResult {
    debug!({
        term_session =? request.terminal_session_id,
        command = request.command,
        cwd = request.working_directory(),
        env =? request.env,
        background = request.background_job,
        pipelined = request.is_pipelined
    }, "Executing command");

    let session_sender = figterm_state
        .with_maybe_id(&request.terminal_session_id.map(FigtermSessionId::new), |session| {
            session.sender.clone()
        });

    if let Some(session_sender) = session_sender {
        let (message, rx) = FigtermCommand::pseudoterminal_execute(
            request.command,
            request.working_directory,
            request.background_job,
            request.is_pipelined,
            request.env,
        );
        session_sender
            .send(message)
            .map_err(|err| format!("failed sending command to figterm: {err}"))?;
        drop(session_sender);

        let response = timeout(Duration::from_secs(10), rx)
            .await
            .map_err(|err| Error::from_std(err).wrap_err("Qterm response timed out after 10 sec"))?
            .map_err(|err| Error::from_std(err).wrap_err("Qterm response failed to receive from sender"))?;

        if let hostbound::response::Response::PseudoterminalExecute(response) = response {
            RequestResult::Ok(Box::new(ServerOriginatedSubMessage::PseudoterminalExecuteResponse(
                PseudoterminalExecuteResponse {
                    stdout: response.stdout,
                    stderr: response.stderr,
                    exit_code: response.exit_code,
                },
            )))
        } else {
            Err("invalid response type".to_string().into())
        }
    } else {
        debug!("executing locally");

        let shell = PlatformState::shell();

        // note: we don't know what shell they use because we don't have any figterm sessions to check
        let args = shell_args(&shell);

        let mut cmd = Command::new(&*shell);
        #[cfg(target_os = "windows")]
        cmd.creation_flags(windows::Win32::System::Threading::DETACHED_PROCESS.0);
        // TODO: better SHELL_ARGs handling here based on shell.
        // TODO: handle wsl distro from FigtermState here.
        cmd.args(args);
        cmd.arg(&request.command);

        if let Some(working_directory) = request.working_directory {
            cmd.current_dir(working_directory);
        }

        set_fig_vars(&mut cmd);

        for EnvironmentVariable { key, value } in &request.env {
            match value {
                Some(value) => cmd.env(key, value),
                None => cmd.env_remove(key),
            };
        }

        let output = cmd
            .output()
            .await
            .map_err(|err| format!("Failed running command {:?}: {err}", request.command))?;

        RequestResult::Ok(Box::new(ServerOriginatedSubMessage::PseudoterminalExecuteResponse(
            PseudoterminalExecuteResponse {
                stdout: String::from_utf8_lossy(&output.stdout).to_string(),
                stderr: if output.stderr.is_empty() {
                    None
                } else {
                    Some(String::from_utf8_lossy(&output.stderr).to_string())
                },
                exit_code: output.status.code(),
            },
        )))
    }
}

pub async fn run(request: RunProcessRequest, state: &FigtermState) -> RequestResult {
    debug!({
        term_session =? request.terminal_session_id,
        exe = request.executable,
        args =? request.arguments,
        cwd = request.working_directory(),
        env =? request.env,
        timeout =? request.timeout,
    }, "Running command");

    let session_sender = state.with_maybe_id(&request.terminal_session_id.map(FigtermSessionId::new), |session| {
        session.sender.clone()
    });

    if let Some(session_sender) = session_sender {
        let (message, rx) = FigtermCommand::run_process(
            request.executable,
            request.arguments,
            request.working_directory,
            request.env,
        );
        session_sender
            .send(message)
            .map_err(|err| format!("failed sending command to figterm: {err}"))?;
        drop(session_sender);

        let timeout_duration = request
            .timeout
            .map(Duration::from)
            .unwrap_or_default()
            .max(Duration::from_secs(10));

        let response = timeout(timeout_duration, rx)
            .await
            .map_err(|_err| "Timed out waiting for figterm response")?
            .map_err(|err| format!("Failed to receive figterm response: {err}"))?;

        if let hostbound::response::Response::RunProcess(response) = response {
            RequestResult::Ok(Box::new(ServerOriginatedSubMessage::RunProcessResponse(
                RunProcessResponse { ..response },
            )))
        } else {
            Err("invalid response type".into())
        }
    } else {
        debug!("running locally");

        // TODO: we can infer shell as above for execute if no executable is provided.
        let mut cmd = Command::new(&request.executable);
        #[cfg(target_os = "windows")]
        cmd.creation_flags(windows::Win32::System::Threading::DETACHED_PROCESS.0);

        if let Some(working_directory) = request.working_directory {
            cmd.current_dir(working_directory);
        } else if let Ok(working_directory) = std::env::current_dir() {
            cmd.current_dir(working_directory);
        }
        for arg in request.arguments {
            cmd.arg(arg);
        }

        set_fig_vars(&mut cmd);

        for var in request.env {
            match var.value {
                Some(val) => cmd.env(var.key, val),
                None => cmd.env_remove(var.key),
            };
        }

        let output = cmd
            .output()
            .await
            .map_err(|err| format!("Failed running command {:?}: {err}", request.executable))?;

        RequestResult::Ok(Box::new(ServerOriginatedSubMessage::RunProcessResponse(
            RunProcessResponse {
                stdout: String::from_utf8_lossy(&output.stdout).to_string(),
                stderr: String::from_utf8_lossy(&output.stderr).to_string(),
                exit_code: output.status.code().unwrap_or(0),
            },
        )))
    }
}

pub async fn write() -> RequestResult {
    RequestResult::error("PseudoterminalWriteRequest is deprecated".to_string())
}
