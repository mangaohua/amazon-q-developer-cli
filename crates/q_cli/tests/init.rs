#[cfg(not(windows))]
use std::io::Write;
#[cfg(not(windows))]
use std::process::{
    Command,
    Stdio,
};

use anstream::println;
#[cfg(not(windows))]
use assert_cmd::prelude::*;
use eyre::Context;
use fig_util::consts::CLI_CRATE_NAME;
use fig_util::consts::build::{
    SKIP_FISH_TESTS,
    SKIP_SHELLCHECK_TESTS,
};
use paste::paste;

macro_rules! init_test {
    ($shell:expr, $stage:expr, $file:expr, [$exe:expr, $($arg:expr),*]) => {
        paste! {
            #[cfg(not(windows))]
            fn [<init_output_ $shell _ $stage _ $file>]() -> Result<String, Box<dyn std::error::Error>> {
                let mut cmd = Command::cargo_bin(CLI_CRATE_NAME)?;
                cmd.arg("init").arg($shell).arg($stage).arg("--rcfile").arg($file);
                cmd.env("Q_INIT_SNAPSHOT_TEST", "1");
                let out = cmd.assert().success().get_output().stdout.clone();
                Ok(String::from_utf8(out)?)
            }

            #[test]
            #[cfg(not(windows))]
            fn [<init_snapshot_ $shell _ $stage _ $file>]() -> Result<(), Box<dyn std::error::Error>> {
                let init = [<init_output_ $shell _ $stage _ $file>]()?;

                insta::assert_snapshot!(init);

                Ok(())
            }

            #[test]
            #[cfg(not(windows))]
            fn [<init_lint_ $shell _ $stage _ $file>]() -> Result<(), Box<dyn std::error::Error>> {
                // Ignore fish post since idk it doesn't work on CI
                if $exe == "fish" && $stage == "post" {
                    return Ok(());
                }

                if $exe == "fish" && SKIP_FISH_TESTS {
                    return Ok(());
                }

                if $exe == "shellcheck" && SKIP_SHELLCHECK_TESTS {
                    return Ok(());
                }

                let init = [<init_output_ $shell _ $stage _ $file>]()?;

                let mut cmd = Command::new($exe);
                cmd$(.arg($arg))*.stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());
                cmd.env("Q_INIT_SNAPSHOT_TEST", "1");

                let child = cmd.spawn().context(format!("{} is not installed", $exe))?;
                write!(child.stdin.as_ref().unwrap(), "{}", init)?;
                let output = child.wait_with_output()?;
                if !output.status.success() {
                    let stdout = String::from_utf8(output.stdout)?;
                    let stderr = String::from_utf8(output.stderr)?;
                    println!("stdout: {stdout}");
                    println!("stderr: {stderr}");

                    // Write shell version to stdout
                    let mut cmd = Command::new($exe);
                    cmd.arg("--version");
                    let out = cmd.output()?;
                    println!("Linter {} version: {}", $exe, String::from_utf8(out.stdout)?);

                    panic!(
                        "linter returned {}. please run `cargo run -p {CLI_CRATE_NAME} -- init {} {} --rcfile {} | {} {}`",
                        output.status, $shell, $stage, $file, $exe, [$($arg),*].join(" ")
                    );
                }

                Ok(())
            }
        }
    };
}

// bash
init_test!("bash", "pre", "bashrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "pre", "bash_profile", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "post", "bashrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "post", "bash_profile", ["shellcheck", "-s", "bash", "-"]);

// zsh
init_test!("zsh", "pre", "zshrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "pre", "zprofile", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "post", "zshrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "post", "zprofile", ["shellcheck", "-s", "bash", "-"]);

// fish
init_test!("fish", "pre", "00_fig_pre", ["fish", "--no-execute"]);
init_test!("fish", "post", "99_fig_post", ["fish", "--no-execute"]);
