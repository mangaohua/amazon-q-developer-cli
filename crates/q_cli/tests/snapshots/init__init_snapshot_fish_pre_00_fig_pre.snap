---
source: q_cli/tests/init.rs
expression: init
---
set -g Q_SHELL "/bin/zsh"
set -g SHOULD_QTERM_LAUNCH "0"
command mkdir -p ~/.local/bin >/dev/null

builtin contains $HOME/.local/bin $PATH
or set --append PATH $HOME/.local/bin

if not test -z "$Q_NEW_SESSION"
    set --erase QTERM_SESSION_ID
    set --erase Q_TERM
    set --erase Q_NEW_SESSION
end

# Load parent from env variables
if test -n "$Q_PARENT"; and test -z "$Q_SET_PARENT"
    set --export Q_PARENT $Q_SET_PARENT
end

if test -z "$SHOULD_QTERM_LAUNCH"
    # 0 = Yes, 1 = No, 2 = Fallback to Q_TERM
    q _ should-figterm-launch 1>/dev/null 2>&1
    set SHOULD_QTERM_LAUNCH $status
end

if test -t 1
    and test -z "$PROCESS_LAUNCHED_BY_Q"
    and command -v qterm 1>/dev/null 2>/dev/null
    and test "$SHOULD_QTERM_LAUNCH" -eq 0 -o \( "$SHOULD_QTERM_LAUNCH" -eq 2 -a \( -z "$Q_TERM" -o \( -z "$Q_TERM_TMUX" -a -n "$TMUX" \) \) \)

    if test -z "$Q_SHELL"
        set Q_SHELL (q _ get-shell)
    end
    set Q_IS_LOGIN_SHELL 0
    if status --is-login
        set Q_IS_LOGIN_SHELL 1
    end

    # Do not launch qterm in non-interactive shells (like VSCode Tasks)
    if status --is-interactive
        set Q_TERM_NAME (command basename "$Q_SHELL")" (qterm)"
        if not set -q Q_TERM_PATH
            if test -x "$HOME/.local/bin/$Q_TERM_NAME"
                set Q_TERM_PATH "$HOME/.local/bin/$Q_TERM_NAME"
            else
                set Q_TERM_PATH (command -v qterm || echo "$HOME/.local/bin/qterm")
            end
        end

        # Need to exec bash because we're using 'exec -a <name>'
        # to set argv[0] and fish's exec doesn't have this option
        exec bash -c "Q_PARENT=$Q_PARENT Q_SHELL=$Q_SHELL Q_IS_LOGIN_SHELL=$Q_IS_LOGIN_SHELL exec -a \"$Q_TERM_NAME\" \"$Q_TERM_PATH\""
    end
end
